File: BOJ/30_최소신장트리/4386_별자리만들기.cpp
#include <iostream>
#include <math.h>
#include <vector>
#include <algorithm>
#define MAXN 100

using namespace std;

int N,par[MAXN+1];
float x,y;
vector<pair<float,pair<int,int> > > tree;
vector<pair<float,float> >stars;

int find(int x) {
    if(par[x] == x) return par[x];
    else return par[x] = find(par[x]);
}

void merge(int x, int y) {
    int p_x = find(x);
    int p_y = find(y);
    if(p_x == p_y) return ;
    par[p_x] = p_y;    
}

float dist(pair<float,float> x, pair<float,float> y) {
    return sqrt(pow(x.first-y.first,2)+pow(x.second-y.second,2));
}

int main() {

    scanf("%d",&N);
    for(int i=0; i<N; i++) {
        par[i] = i;
        scanf("%f %f",&x,&y);
        stars.push_back(make_pair(x,y));
    }
    for(int i=0; i<N; i++) {
        for(int j=i+1; j<N; j++) {
            tree.push_back(make_pair(dist(stars[i],stars[j]),make_pair(i,j)));
        }
    }
    sort(tree.begin(),tree.end());
    float cost = 0;
    for(int i=0; i<tree.size(); i++) {
        float c = tree[i].first;
        int a = tree[i].second.first;
        int b = tree[i].second.second;
        if(find(a) == find(b)) continue;
        merge(a,b);
        cost += c;
    }
    printf("%.2f\n",cost);

    return 0;
}

File: BOJ/30_최소신장트리/4386_별자리만들기.py
import sys
import math

read = sys.stdin.readline

def find(x) :
    if par[x] == x : return par[x]
    else :
        par[x] = find(par[x])
        return par[x]

def merge(x,y) :
    p_x = find(x)
    p_y = find(y)

    if p_x == p_y : return
    par[p_x] = p_y

def dist(x,y) :
    return math.sqrt((x[0]-y[0])**2 + (x[1]-y[1])**2)

N = int(read())
pt = [tuple(map(float,read().split())) for _ in range(N)]
par = [i for i in range(N)]
tree = []
for i in range(N) :
    for j in range(i+1,N) :
        tree.append([dist(pt[i],pt[j]),i,j])

tree.sort()
cost = 0
for c,a,b in tree :
    if find(a) == find(b) : continue
    merge(a, b)
    cost += c
print(round(cost,2))
